word_segmentation.py

```diff
@@ -0,0 +1,709 @@
+        st_ind += 1
+    return new_line
+
+
+def clean_line(line):
+    """
+    This line cleans a line as follows such that it is ready for process by different components of the code. It returns
+    the clean line or -1, if the line should be omitted.
+        1) remove tags and https from the line.
+        2) Put a | at the begining and end of the line if it isn't already there
+        3) if line is very short (len < 3) or if it is all in English or it has a link in it, return -1
+    """
+
+    # Remove lines with links
+    if "http" in line or len(line) == 0:
+        return -1
+
+    # Remove texts between following tags
+    line = remove_tags(line, "<NE>", "</NE>")
+    line = remove_tags(line, "<AB>", "</AB>")
+
+    # Remove lines that are all fully in English
+    if is_all_english(line):
+        return -1
+
+    # Exclude cases such as "", " ", "| ", " |", etc.
+    if len(line) < 3:
+        return -1
+
+    # Add "|" to the end of each line if it is not there
+    if len(line) >= 1 and line[len(line) - 1] != '|':
+        line += "|"
+
+    # Adding "|" to the start of each line if it is not there
+    if line[0] != '|':
+        line = '|' + line
+
+    return line
+
+
+def preprocess():
+    """
+    This function uses the BEST data set to
+        1) compute the grapheme cluster dictionary that holds the frequency of different grapheme clusters
+        2) demonstrate the performance of icu word breakIterator and compute its accuracy
+    """
+    grapheme_clusters_dic = dict()
+    icu_mismatch = 0
+    icu_total_bies_lengths = 0
+    for cat in ["news", "encyclopedia", "article", "novel"]:
+        for text_num in range(1, 96):
+            text_num_str = "{}".format(text_num).zfill(5)
+            file = open("./Data/Best/{}/{}_".format(cat, cat) + text_num_str + ".txt", 'r')
+            line_counter = 0
+            while True:
+                line = file.readline().strip()
+                if not line:
+                    break
+
+                line = clean_line(line)
+                if line == -1:
+                    continue
+
+                # Finding word breakpoints using the segmented data
+                word_brkpoints = []
+                found_bars = 0
+                for i in range(len(line)):
+                    if line[i] == '|':
+                        word_brkpoints.append(i - found_bars)
+                        found_bars += 1
+
+                # Creating the unsegmented line
+                unsegmented_line = line.replace("|", "")
+
+                # Making the grapheme clusters brkpoints
+                chars_break_iterator = BreakIterator.createCharacterInstance(Locale.getUS())
+                chars_break_iterator.setText(unsegmented_line)
+                char_brkpoints = [0]
+                for brkpoint in chars_break_iterator:
+                    char_brkpoints.append(brkpoint)
+
+                # Storing the grapheme clusters and their frequency in the dictionary
```

Comment:
You can use a Python Counter for this.

https://docs.python.org/3.8/library/collections.html#collections.Counter

Reply:
Yes, if I want to do that I can store all grapheme clusters in a list and then at the end, use `Counter` to get a dictionary from that list. I don't see the point of this though, why I don't make the dictionary directly along the way? This way I don't need to store an extra list too.

Reply:
Another option is `defaultdict`.

https://docs.python.org/3.10/library/collections.html#collections.defaultdict